README


This archive contains the distribution Boost-Geometry-Utils,
version 0.06:

  Bindings for the Boost Geometry library

This software is copyright (c) 2013 by Alessandro Ranellucci.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.




CHANGES
==================================================
Changes from 2012-01-28 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.06 at 2013-01-27 23:25:27 +0000
-----------------------------------------

  Change: 29af743cedd602af65deecf056b59ac0c909282a
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-28 00:25:03 +0000

    Update POD 

  Change: 58736a034fb398d7e4f6a00b975a14ce2d122d7a
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-27 23:58:02 +0000

    Fixed arrayref to polygon conversion and added unit test 

  Change: 885c60668125cd281b7783baf7dc69951e0d0f2e
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-27 16:30:38 +0000

    Finished polygon2perl() and added unit test for conversion roundtrip 

  Change: 1df409ec73fb4de93db234f9634399c64aba0314
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-27 15:08:11 +0000

    Update copyright year 

  Change: 0a0b9442fb84af318e888a0064da6f54474fc9ae
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2013-01-27 15:07:56 +0000

    Added a couple conversion methods to help debugging 

  Change: c3fc121bf240a9b15b69d356b4ffc8088ed52609
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-11-01 17:18:17 +0000

    Use integers instead of doubles and fix test to use integers. 

  Change: 71ce53ba110bea82bb32c63e8aef82719bcbea80
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-10 22:28:25 +0000

    Convert polygon() and linestring() functions to use non-wkt
    conversion. 

  Change: f95918786093c58de80ab62c0f47355c668fb48e
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-08 11:41:27 +0000

    Attempt to fix/workaround leaking memory. 

  Change: fec904e6630837aef63260b5bdde034b6d570ae5
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-08 08:30:55 +0000

    Remove pointless reference to boost::geometry::append. 

  Change: e917ff4dbc3eeaa3e2e51cacb0dad649712310d1
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-08 07:50:09 +0000

    Add point_covered_by_polygon and add tests to cover "edge" cases. 

  Change: 259eed2f7952443be651f87d50772181a363c105
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-07 23:40:35 +0000

    Add point_within_polygon and linestring_simplify functions and tests.

    There is a bug in point_within_polygon triggered by:

    point_within_polygon($point_in_hole, [$hole_in_square,[]])

    which I've avoided by changing the test. This needs fixing. 

  Change: d93104e6ebb7ed61f6dca2bd660aca10e6343783
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-07 22:38:18 +0000

    Fix typo that accidentally worked. 

  Change: ffd640666a4dfc64eaa481e701e5fa10456a0b03
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-07 22:19:50 +0000

    Trivial whitespace change. 

  Change: 07b52f3ca576db64b4488630b5234e18e7687b84
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-07 22:19:17 +0000

    Put back 64-bit int support. 

  Change: ce6abfcd8966a3873b5f8c0469f56ac34183bfa8
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-07 22:06:13 +0000

    Support old API and new polygon_multi_linestring_intersection
    function.

    Old API is the same as before but the new version does the return
    value conversion using the new XS code path rather than the old wkt
    code. 

  Change: a36005753777bbe951050c1c6f6b1fdb289063ab
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-07 20:55:10 +0000

    Refactor add_outer and add_hole to single add_ring function. 

  Change: 3813ed19a4aae58ccaa3b4b2c1aeb58e68aea791
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-07 20:30:56 +0000

    Remove some debug and add macros to simplify code. 

  Change: ab1b86acb5ce159af54e23c58aed54768ef0ed31
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-07 18:09:58 +0000

    Implement boost to perl conversion for multi_linestring. 

  Change: 9808ba60cc64a6dbaddde6e98493e5c8ea7cf5b9
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-06 23:11:15 +0000

    Convert perl to multi_linestring. 

  Change: 641f20262b6a0fa246b525f3f1639062bc71473a
  Author: Mark Hindess <soft-github@temporalanomaly.com>
  Date : 2012-07-06 19:14:09 +0000

    Fix perl to boost polygon conversion.

    Lots of debug left in and lots of cleanup to do but it works. 

  Change: d98d2ecd5377635b1c29e947d7e44a82550ac634
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2012-07-03 19:10:28 +0000

    Unfinished draft for avoiding the WKT roundtrip and having XS work
    directly on Perl data structures 

-----------------------------------------
version 0.05 at 2012-04-10 08:39:45 +0000
-----------------------------------------

  Change: 680789acf635cfcbdfd5a8c23de5ca99a7ec8680
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2012-04-10 10:39:45 +0000

    Added build and configure modules to dist.ini and thus META.yml 

-----------------------------------------
version 0.04 at 2012-04-09 19:51:56 +0000
-----------------------------------------

  Change: 3f86a958746db1b72bf49e12e042c87f226089c2
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2012-04-09 21:51:56 +0000

    Put XSLoader::log() outside of BEGIN in order to allow for DZ to add
    $VERSION 

-----------------------------------------
version 0.03 at 2012-04-09 19:41:18 +0000
-----------------------------------------

  Change: 1bdc8c6b368bb403f98a5b9b0b55821f133aff91
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2012-04-09 21:41:18 +0000

    Remove $VERSION so that $Dist::Zilla manages it 

  Change: 3563c5b03dc794210339f785f36cebb02eda22d0
  Author: Alessandro Ranellucci <aar@cpan.org>
  Date : 2012-04-09 21:36:55 +0000

Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Bindings for the Boost Geometry library",
  "AUTHOR" => "Alessandro Ranellucci <aar\@cpan.org>",
  "BUILD_REQUIRES" => {
    "Module::Build" => "0.3601",
    "Test::More" => 0
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30",
    "Module::Build" => "0.3601"
  },
  "DISTNAME" => "Boost-Geometry-Utils",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Boost::Geometry::Utils",
  "PREREQ_PM" => {},
  "VERSION" => "0.02",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/Boost/Geometry/Utils.pm
package Boost::Geometry::Utils;
{
  $Boost::Geometry::Utils::VERSION = '0.06';
}
# ABSTRACT: Bindings for the Boost Geometry library
use strict;
use warnings;

require Exporter;
our @ISA = qw(Exporter);

use XSLoader;
XSLoader::load('Boost::Geometry::Utils', $Boost::Geometry::Utils::VERSION);

our @EXPORT_OK = qw(polygon_to_wkt linestring_to_wkt wkt_to_multilinestring
    polygon linestring polygon_linestring_intersection
    polygon_multi_linestring_intersection
    point_within_polygon point_covered_by_polygon linestring_simplify);

sub polygon_to_wkt {
    sprintf 'POLYGON(%s)', join ',', map { sprintf '(%s)', join ',', map { join ' ', @$_ } @$_ } @_;
}

sub linestring_to_wkt {
    sprintf "MULTILINESTRING(%s)", join ',', map { sprintf '(%s)', join ',', map { join ' ', @$_ } @$_ } @_;
}

sub wkt_to_multilinestring {
    return [] if $_[0] eq 'MULTILINESTRING()';
    $_[0] =~ s/^MULTILINESTRING\(\(//;
    $_[0] =~ s/\)\)$//;
    [ map [ map [ split / / ], split /,/ ], split /\),\(/, $_[0] ];
}

sub polygon {
    _polygon(\@_);
}

sub linestring {
    _multi_linestring(\@_)
}

sub multi_linestring {
    _multi_linestring(\@_)
}

1;


__END__
=pod

=head1 NAME

Boost::Geometry::Utils - Bindings for the Boost Geometry library

=head1 VERSION

version 0.06

=head1 SYNOPSIS

    use Boost::Geometry::Utils qw(polygon linestring polygon_linestring_intersection);
    
    my $square = [  # ccw
        [10, 10],
        [20, 10],
        [20, 20],
        [10, 20],
    ];
    my $hole_in_square = [  # cw
        [14, 14],
        [14, 16],
        [16, 16],
        [16, 14],
    ];
    my $polygon = polygon($square, $hole_in_square);
    my $linestring = linestring([ [5, 15], [30, 15] ]);
    
    my $intersection = polygon_linestring_intersection($polygon, $linestring);
    
    # $intersection is:
    # [
    #     [ [10, 15], [14, 15] ],
    #     [ [16, 15], [20, 15] ],
    # ]

=head1 ABSTRACT

This module provides bindings to perform some geometric operations using
the Boost Geometry library. It does not aim at providing full bindings
for such library, and that's why I left the I<Boost::Geometry> namespace
free. I'm unsure about the optimal architectural for providing full 
bindings, but I'm interested in such a project -- so, if you have ideas
please get in touch with me.

B<Warning:> the API could change in the future.

=head1 METHODS

=head2 polygon

Converts one or more arrayref(s) of points to a Boost Geometry polygon 
structure. Note that a Boost Geometry polygon is actually a polygon
with holes.

=head2 linestring

Converts an arrayref of points to a Boost Geometry linestring.

=head2 multi_linestring

Converts an arrayref of arrayrefs of points to a Boost Geometry multilinestring.

=head2 polygon_linestring_intersection

Performs an intersection between the supplied polygon and linestring,
and returns an arrayref of linestrings (represented as arrayrefs of
points).
Note that such an intersection is also called I<clipping>.

=head2 polygon_multi_linestring_intersection

Same as I<polygon_linestring_intersection> but it accepts a multilinestring
object to perform multiple clippings in a single batch.

=head2 polygon_to_wkt

Converts one or more arrayref(s) of points to a WKT representation of
a polygon (with holes).

=head2 linestring_to_wkt

Converts an arrayref of points to a WKT representation of a multilinestring.

=head2 wkt_to_multilinestring

Parses a MULTILINESTRING back to a Perl data structure.

=head2 linestring_simplify

Accepts an arrayref of points representing a linestring and a numeric tolerance 
and returns an arrayref of points representing the simplified linestring.

=head2 point_covered_by_polygon

Accepts a point and an arrayref of points representing a polygon and returns true 
or false according to the 'cover_by' strategy.

=head2 point_within_polygon

Accepts a point and an arrayref of points representing a polygon and returns true 
or false according to the 'within' strategy.

=head1 ACKNOWLEDGEMENTS

Thanks to mauke and mst (Matt S. Trout (cpan:MSTROUT) <mst@shadowcat.co.uk>)
for their valuable help in getting this to compile under Windows (MinGW) too.
Thanks to Mark Hindness for his work on data types conversion.

=head1 AUTHOR

Alessandro Ranellucci <aar@cpan.org>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Alessandro Ranellucci.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

